### PR TITLE
[WIP] Lint script for checking driver compliance to the base / standard Libcloud API and quality checking

### PR DESCRIPTION
**Pull request by Kami**
This pull request works on the great initial work and idea from @Jc2k (#211).
## Background

I've talked with @tonybaloney the other day on how we could make the library better and more user-friendly. Usually the biggest problem and "unfriendliness" is when the driver doesn't fully comply to the base API. From a high level perspective those small things (e.g. argument is missing an `ex_` prefix, etc.) don't seen like a big deal, but they actually are (death by the thousand cuts).

This causes problems because it makes building truly provider agnostic applications harder and means that user needs to build another layer on top of Libcloud which handles those differences and this makes the library less attracting (hey, Libcloud is supposed to abstract those differences away!).

As of now, we do this "compliance against the base / standard API" checking manually during code reviews. Obviously manual process is error prone and wastes a tons of time which could be spend on reviewing things which actually add value.

I think the first step for improving the quality is having a script and automated check (linter or whatever you might call it) which runs on every commit and reports all the violations.
## TODO
- [ ] Finish it up (still tons of work and refactoring to be done)
- [ ] Hook it up to the CI (start with the existing warnings not being fatal)
- [ ] Start fixing and improving the existing code (not part of this PR)

### CODE REVIEW

File: contrib/lint.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+from collections import defaultdict
+
+import libcloud
+
+from libcloud.compute.providers import get_driver as get_compute_driver
+from libcloud.compute.base import NodeDriver
+
+import libcloud.dns.providers
+from libcloud.dns.base import DNSDriver
```
Comment (tonybaloney) [inline contrib/lint.py:42]: Add `# pylint: disable=unused-import` to lines 42, 45 and 48

File: contrib/lint.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+        extra_args = (driver_args - core_args)
+        for extra in extra_args:
+            if not extra.startswith('ex_'):
+                message = "Core API function shouldn't take arg '%s'. Should it be prefixed with ex_?" % extra
+                warning = get_warning_object(obj=method, message=message)
+                warnings.append(warning)
+
+    # Filter out empty warning objects (dupes)
+    warnings = [warning for warning in warnings if warning is not None]
```
Comment (tonybaloney) [inline contrib/lint.py:170]: call the temp var (`warning`) something else because it conflicts with the previous scope

**Comment by tonybaloney** [comment]
@Kami are you going to merge this and then keep the fixes to the drivers seperate?

**Comment by Kami** [comment]
The script / code itself still needs quite a bit more of (refactoring) love before I plan to merge it.

I mostly just wanted to get some feedback on the overall approach to begin with :)

**Comment by Kami** [comment]
Old and out of date PR. Closing.

**Pull request closed without merging**